### PR TITLE
Produce proper error message if a matching typedesc function cannot b…

### LINKED ISSUE #9961: internal error when assigning a type to a constant of typedesc
The following crashes the compiler:

```nim
import typetraits
import tables

proc test(v: typedesc) =
  echo v.type.name

# This works:
test Table

# This also works:
const a = Table
test a

# But this crashes the compiler
const b: typedesc = Table
test b
```

```
internal error: (filename: "semmagic.nim", line: 162, column: 17)
```

```
Nim Compiler Version 0.19.0 [Linux: amd64]
Compiled at 2018-09-26
Copyright (c) 2006-2018 by Andreas Rumpf

git hash: f6c5c636bb1a1f4e1301ae0ba5a8afecef439132
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by deech**
…e found.

Fixes #9961.

### CODE REVIEW

File: tests/metatype/tmetatype_issues.nim [hunk 1]
```diff
@@ -156,3 +156,11 @@ block t3338:
   var t2 = Bar[int32]()
   t2.add()
   doAssert t2.x == 5
+
+block t9961:
+  proc n(t:typedesc) =
+    echo t.type.name
+  const t : typedesc = int
+  assert(not
+         compiles((block:
```
Comment (timotheecour) [inline tests/metatype/tmetatype_issues.nim]: /cc @araq 
that'd be a perfect, typical use case for `compilesWithContext` I suggested here: https://github.com/nim-lang/Nim/issues/8803
currently the error message isn't being tested for, and PR author @deech would have to run through significant contorsions to test for it (moving this to a separate test file for example). I run into this all the time. I'm willing to try implementing it if feature is pre-approved.


here, its use would look like:

```nim
let context = compilesWithContext(block: (n t))
doAssert: not context.ok
doAssert: context. errorMsg.startsWith("Could not find a matching ...")
```

Comment (Araq) [reply tests/metatype/tmetatype_issues.nim]: Again, error messages can be checked for ``cmd: "nim check $file"``. Again, ``compilesWithContext`` is the worst hack in the history of mankind and won't be accepted.

File: compiler/semcall.nim [hunk 2]
```diff
@@ -502,6 +502,21 @@ proc semOverloadedCall(c: PContext, n, nOrig: PNode,
                        filter: TSymKinds, flags: TExprFlags): PNode =
   var errors: CandidateErrors = @[] # if efExplain in flags: @[] else: nil
   var r = resolveOverloads(c, n, nOrig, filter, flags, errors, efExplain in flags)
+
+  if r.state == csMatch and r.exactMatches == 0 and r.fauxMatch == tyFromExpr:
+    let diagnostic =
+      "Could not find a matching '" &
+        r.calleeSym.name.s &
+        "' for this typedesc because it does not carry enough information to resolve to a concrete type. " &
```
[the remark this answers is not in the document]
Comment (deech) [reply compiler/semcall.nim]: Is this a big deal? It's only two lines.

[the remark this answers is not in the document]
Comment (timotheecour) [reply compiler/semcall.nim]: not a big deal; consider using it next time; `"foo" & "\nbar"` is from C-era

File: compiler/semcall.nim [hunk 3]
```diff
@@ -502,6 +502,21 @@ proc semOverloadedCall(c: PContext, n, nOrig: PNode,
                        filter: TSymKinds, flags: TExprFlags): PNode =
   var errors: CandidateErrors = @[] # if efExplain in flags: @[] else: nil
   var r = resolveOverloads(c, n, nOrig, filter, flags, errors, efExplain in flags)
+
+  if r.state == csMatch and r.exactMatches == 0 and r.fauxMatch == tyFromExpr:
```
Comment (Araq) [inline compiler/semcall.nim]: Arbitrary logic. Why ``exactMatches == 0``? What if there are subtype matches?

**Comment by Araq** [comment]
No feedback, closing.

**Pull request closed without merging**